### ISSUE TITLE
Increasing performance

### ISSUE THREAD

**Issue by swifmaneum** [lantunes/cellpylib#3]
Hi there and thanks for the intuitive library!

I'd like to recreate the experiments from the paper you mention in the readme: "Evolving Cellular Automata with Genetic Algorithms: A Review of Recent Work" from Melanie Mitchell, James P. Crutchfield, and Rajarshi Das. However, evolving a population of say 100 automata for some generations takes quite a while.

Profiling showed one simple way to reduce the runtime: In `binary_rule` the rule is given as an int and converted to bits using the `int_to_bits` method in every call. Storing the rule as bits and skipping the conversion reduced the runtime of my experiments. Is there a way to support this in the library? I'm thinking of a possibility to simply hand a binary np-array into the `binary_rule` method, see exemplary code below.

The same goes with the `bits_to_int` method. After profiling with above code this is the method causing the most runtime. A possible performance improvement would be to calculate the powers of 2 beforhand and then only compute the dot product of neighbourhood and the powers of 2 as suggested [here](https://stackoverflow.com/a/63130015/5730444).

I've sketched both perfomance improvements in a `fast_binary_rule` method:

```
def fast_binary_rule(neighbourhood, binary_rule, powers_of_two, scheme=None):
    state_int = neighbourhood.dot(powers_of_two)
    n = 2 ** len(neighbourhood)
    if scheme == 'nks':
        return binary_rule[(n - 1) - state_int]
    return binary_rule[state_int]
```

The usage is pretty much the same but with the additional `powers_of_two` parameter. Here a full example:
```
import cellpylib as cpl
import numpy as np


def fast_binary_rule(neighbourhood, binary_rule, powers_of_two, scheme=None):
    state_int = neighbourhood.dot(powers_of_two)
    n = 2 ** len(neighbourhood)
    if scheme == 'nks':
        return binary_rule[(n - 1) - state_int]
    return binary_rule[state_int]


timesteps = lattice_size = 49
neighbourhood_radius = 3
powers_of_two = a = 2 ** np.arange(neighbourhood_radius * 2 + 1)[::-1]

cellular_automaton = cpl.init_random(lattice_size)

rule = np.random.choice([0, 1], size=(128,))

cellular_automaton = cpl.evolve(cellular_automaton, timesteps=timesteps,
                                apply_rule=lambda n, c, t: fast_binary_rule(n, rule, powers_of_two),
                                r=neighbourhood_radius)

cpl.plot(cellular_automaton)
```

On my machine using the `fast_binary_rule` cuts the execution time of my experiments roughly in half. Would you be interessted in integrating these performance improvements?

And: Are there other performance improvements to further reduce runtime?

**Comment by lantunes** [comment]
Hi, thanks for the message, and for the suggestions! My primary goal with the project was to achieve breadth and expressiveness, and this sometimes comes at the cost of longer running times. However, there are definitely opportunities for making the code more efficient. If you can put all your changes in a pull request, I can certainly merge them. There is a suite of tests, and they must all pass. 

In terms of other performance improvements, I can't say that I have given it much thought, as I have limited time to devote to this project these days. That being said, I have another project, https://github.com/lantunes/netomaton, that also supports CA, and provides the option to process the evolution of the system in parallel. Depending on the size of the system, and the availability of processor cores, parallel processing may also speed things up.

**Comment by swifmaneum** [comment]
You're right. I tried to come up with a solution that fits the style of the project, see #4. 

This should not break the API and leaves all test intact. What do you think? I might add some test cases later if you care. 

I like the netomaton project. I might try it out later and compare the performance ;)

**Comment by lantunes** [comment]
Thanks for the pull request! I will review it. I will also suggest some test cases.

Another point about performance: in my experience, supplying a lambda as the `apply_rule` argument may be slower than defining a function and providing the reference to the function as the argument, though I may be wrong. 

For example:
```
def my_rule(n, c, t):
    return cpl.binary_rule(n, 30, scheme="nks")

cpl.evolve(cpl.init_simple(size=100), timesteps=100, apply_rule=my_rule, r=1)
```

**Comment by lantunes** [comment]
closed by #4
